streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java

```diff
@@ -564,12 +609,84 @@ public void shouldSetDifferentDefaultsIfEosEnabled() {
         assertTrue((Boolean) producerConfigs.get(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG));
         assertThat(producerConfigs.get(ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG), equalTo(Integer.MAX_VALUE));
         assertThat(streamsConfig.getLong(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG), equalTo(100L));
+
+        return producerConfigs;
     }
 
     @Test
-    public void shouldNotOverrideUserConfigRetriesIfExactlyOnceEnabled() {
-        final int numberOfRetries = 42;
+    public void shouldOverrideUserConfigTransactionalIdIfEosAlphaEnabled() {
         props.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, EXACTLY_ONCE);
+        props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "user-TxId");
+        final StreamsConfig streamsConfig = new StreamsConfig(props);
+
+        final Map<String, Object> producerConfigs = streamsConfig.getProducerConfigs(clientId);
+
+        assertThat(producerConfigs.get(ProducerConfig.TRANSACTIONAL_ID_CONFIG), is(nullValue()));
+    }
+
+    @Test
+    public void shouldOverrideUserConfigTransactionalIdIfEosBetaEnabled() {
+        props.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, EXACTLY_ONCE_BETA);
+        props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "user-TxId");
+        final StreamsConfig streamsConfig = new StreamsConfig(props);
+
+        final Map<String, Object> producerConfigs = streamsConfig.getProducerConfigs(clientId);
+
+        assertThat((String) producerConfigs.get(ProducerConfig.TRANSACTIONAL_ID_CONFIG), startsWith(applicationId + "-"));
+    }
+
+    @Test
+    public void shouldSetTransactionalIdIfEosBetaEnabledWhileUpgradingFromPreEosVersion() {
+        props.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, EXACTLY_ONCE_BETA);
+
+        for (final String upgradeFrom : mkSet(
+            StreamsConfig.UPGRADE_FROM_0100,
+            StreamsConfig.UPGRADE_FROM_0101,
+            StreamsConfig.UPGRADE_FROM_0102)) {
```

Comment:
If we are upgrading from a non-eos version to eos-beta, there is no need to do producer-per-task at all, and thus, we switch to producer per-thread directly. Strictly speaking, a two rolling bounce upgrade would be even be required for the eos-beta opt-in, but people might need to do a two rolling bounce upgrade for other reasons. (Similar pattern for other test)